[GLIB] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE and IGNORE_CLANG_WARNINGS from NetworkCacheDataGLib

csaavedra opened the pull request:
#### 31f2a10dbe959b4849271e70f2c7d98b68dcc620
<pre>
[GLIB] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE and IGNORE_CLANG_WARNINGS from NetworkCacheDataGLib
<a href="https://bugs.webkit.org/show_bug.cgi?id=304742">https://bugs.webkit.org/show_bug.cgi?id=304742</a>

Reviewed by Chris Dumez.

* Source/WebKit/NetworkProcess/cache/NetworkCacheDataGLib.cpp:
(WebKit::NetworkCache::Data::Data):
(WebKit::NetworkCache::concatenate):

Canonical link: <a href="https://commits.webkit.org/304981@main">https://commits.webkit.org/304981@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ef2f7a02cec8fc9d4ab6c3ca76400b6b0d1bf0e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137088 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9448 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48375 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144831 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90061 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10151 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9575 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104832 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140033 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7462 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122846 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85667 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7099 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4805 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5420 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116447 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147587 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9131 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41579 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113186 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9149 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7674 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113516 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28831 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7022 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119109 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63469 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9179 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37163 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8903 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72745 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9120 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8972 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheDataGLib.cpp:
```diff
@@ -47,9 +47,7 @@ namespace NetworkCache {
 Data::Data(std::span<const uint8_t> data)
 {
     uint8_t* copiedData = static_cast<uint8_t*>(fastMalloc(data.size()));
```
This is OK but you could avoid calling `unsafeMakeSpan()` also if you used `MallocSpan<uint8_t>::malloc()` instead of `fastMalloc()`. Then you can get the `mutableSpan()` from the MallocSpan to do the memcpySpan(). You'd likely have to call `leakSpan()` at the end to call `g_bytes_new_with_free_func()`. Might be worth considering a wrapper for `g_bytes_new_with_free_func` that takes spans though not truly required.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheDataGLib.cpp:
```diff
@@ -106,15 +104,9 @@ Data concatenate(const Data& a, const Data& b)
-    const auto* bData = g_bytes_get_data(b.bytes(), &bLength);
-    memcpy(data + aLength, bData, bLength);
-    WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+    memcpySpan(unsafeMakeSpan(data, a.size()), a.span());
```
See, I think the `unsafeMakeSpan()` here makes the code look unsafe because the size is not even the size we malloc'd. There is a risk of memory safety still due to user error.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheDataGLib.cpp:
```diff
@@ -106,15 +106,10 @@ Data concatenate(const Data& a, const Data& b)
-    memcpy(data + aLength, bData, bLength);
-    WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+    auto span = MallocSpan<uint8_t>::malloc(size);
+    memcpySpan(span.mutableSpan().first(a.size()), a.span());
```
`.first(a.size())` is not needed.

cdumez reviewed (approved):
LGTM with nit.

csaavedra committed:
[GLIB] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE and IGNORE_CLANG_WARNINGS from NetworkCacheDataGLib
https://bugs.webkit.org/show_bug.cgi?id=304742

Reviewed by Chris Dumez.

* Source/WebKit/NetworkProcess/cache/NetworkCacheDataGLib.cpp:
(WebKit::NetworkCache::Data::Data):
(WebKit::NetworkCache::concatenate):

Canonical link: https://commits.webkit.org/304981@main
